File: Data Structures & Algorithms/binary-search/submission-1.swift
class Solution {
    func search(_ nums: [Int], _ target: Int) -> Int {
        var L = 0
        var R = nums.count - 1
        
        while L <= R {
            // middle
            let mid = (L + R) / 2
            
            if target > nums[mid] {
                // left must increase
                L = mid + 1
            }else if target < nums[mid]{
                R = mid - 1
            }
            else{
                return mid
            }
            
        }
        return -1
    
    }
}

File: Data Structures & Algorithms/subarray-product-less-than-k/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number}
     */
    numSubarrayProductLessThanK(nums, k) {
        if (k <= 1) return 0;

        let left = 0;
        let product = 1;
        let count = 0;

        for (let right = 0; right < nums.length; right++) {
            product *= nums[right];

            // shrink window until product < k
            while (product >= k) {
                product /= nums[left];
                left++;
            }

            // number of valid subarrays ending at `right`
            count += (right - left + 1);
        }

        return count;
    }
}
